Add a general "getting started" document for WC Blocks extensibility (#53899)
woocommerce/woocommerce: docs/cart-and-checkout-blocks/extensibility-getting-started.md

---
post_title: Cart and Checkout - Extensibility getting started guide
menu_title: Extensibility
tags: reference
---
<!-- markdownlint-disable MD041 -->

This document is a high-level overview of the moving parts required to extend the Cart and Checkout blocks.

To get started, it is recommended to first read the [Block Development Environment](https://developer.wordpress.org/block-editor/getting-started/devenv/) documentation from WordPress and follow [Tutorial: Build your first block
](https://developer.wordpress.org/block-editor/getting-started/tutorial/).

## Example block template package

There is an example block template in the WooCommerce repository. Having this template set up while reading this document may help to understand some of the concepts discussed. See the [`@woocommerce/extend-cart-checkout-block` package documentation](https://github.com/woocommerce/woocommerce/tree/trunk/packages/js/extend-cart-checkout-block#woocommerceextend-cart-checkout-block) for how to install and run the example block.

> (Note: the code in the repository linked above will not be very useful alone, the code there is templating code. It will be transformed into regular JS and PHP after following the README instructions.)

Reviewer: ```suggestion
(Note: the code in the repository linked above will not be very useful alone; the code there is templating code. It will be transformed into regular JS and PHP after following the README instructions.)
```

> (Note: the code in the repository linked above will not be very useful alone, the code there is templating code. It will be transformed into regular JS and PHP after following the README instructions.)

Reviewer: Make README a link.

  : [
    'dependencies' => [],
    'version'      => $this->get_file_version( $script_path ),
  ];

wp_register_script(
  'example-blocks-integration-handle',
  $script_url,
  $script_asset['dependencies'],
  $script_asset['version'],
  true
);
```

Please see the [Cart and Checkout – Handling scripts, styles, and data](https://developer.woocommerce.com/docs/cart-and-checkout-handling-scripts-styles-and-data/) document for more information about how to correctly register scripts using the `IntegrationInterface`.

### Creating a block

In the example block, there is a "checkout-newsletter-subscription-block" directory which contains the files needed to register an inner block in the Checkout. The example block template is only set up to import and build a single block, but the Webpack config can be modified to build multiple blocks. Doing this is not supported as part of this document, refer to the [Webpack documentation](https://webpack.js.org/concepts/) instead.

The principles covered in [Tutorial: Build your first block
](https://developer.wordpress.org/block-editor/getting-started/tutorial/) apply here too.

### Modifying existing values on the front-end

> You may not need to create a block to get your extension working the way you want, for example if your extension only modifies existing content through filters.

Reviewer: ```suggestion
You may not need to create a block to get your extension working the way you want, for example, if your extension only modifies existing content through filters.
```

### Importing WooCommerce components into your extension

Components can be imported from `@woocommerce/blocks-components` (externalized to `window.wc.blocksComponents` by `@woocommerce/dependency-extraction-webpack-plugin`). The list of available components can be seen in [the WooCommerce Storybook](https://woocommerce.github.io/woocommerce/?path=/docs/woocommerce-blocks_external-components-button--docs), listed under "WooCommerce Blocks -> External components".

An example of importing the `Button` component is:

```js
import { Button } from '@woocommerce/blocks-components';

const MyComponent = () => {
  return <div class="my-wrapper">
    <Button type="button" />
  </div>;
}
```

### Importing WooCommerce (React) hooks

Currently, none of our hooks are designed to be used externally, so trying to import hooks such as `useStoreCart` is not supported. Instead, getting the data from the [`wc/store/...`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/docs/third-party-developers/extensibility/data-store/) data stores is preferred.

## Back-end extensibility

### Modifying information during the Checkout process

> Modifying the way the server-side part of the Cart and Checkout blocks works is possible by using PHP. Some actions and filters from the shortcode cart/checkout experience will work, however some do not, we do have a working document ([Hook alternatives document](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce-blocks/docs/third-party-developers/extensibility/hooks/hook-alternatives.md)) that outlines which hooks are supported, and which are not as well as listing any alternative approaches.

Reviewer: ```suggestion
Modifying the server-side part of the Cart and Checkout blocks is possible using PHP. Some actions and filters from the shortcode cart/checkout experience will work too, but not all of them. We have a working document ([Hook alternatives document](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce-blocks/docs/third-party-developers/extensibility/hooks/hook-alternatives.md)) that outlines which hooks are supported as well as alternatives.
```

### Extending Store API

> If you need to extend the way Store API works and the data in the responses the best documentation is the [Extending the Store API](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce-blocks/docs/third-party-developers/extensibility/rest-api).

Reviewer: ```suggestion
If you need to change how the Store API works, or extend the data in the responses, see [Extending the Store API](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce-blocks/docs/third-party-developers/extensibility/rest-api).
```
